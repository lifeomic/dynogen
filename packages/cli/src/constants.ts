export const DEFAULT_DYNOGEN_CONFIG_PATH = 'dynogen.yml';
