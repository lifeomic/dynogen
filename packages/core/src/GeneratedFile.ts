import * as joi from '@hapi/joi';

export interface GeneratedFileConfig {
  readonly outPath: string;
  readonly overwrite?: boolean;
}

export const GeneratedFileConfigSchema = joi.object({
  outPath: joi.string().required(),
  overwrite: joi.bool()
});
