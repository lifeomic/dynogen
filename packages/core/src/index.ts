export * from './generate';
export * from './config';
export * from './File';
export * from './Mapper';
export * from './Item';
