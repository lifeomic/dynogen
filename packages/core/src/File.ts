import { GENERATED_FILE_HEADER_COMMENT } from './constants';

export class File {
  constructor(public readonly path: string, content: string) {
    this.content = `${GENERATED_FILE_HEADER_COMMENT}\n${content}`;
  }

  public readonly content: string;
}
