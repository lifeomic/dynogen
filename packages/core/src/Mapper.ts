import * as path from 'path';
import * as ejs from 'ejs';
import * as joi from '@hapi/joi';
import { JSONSchema4TypeName } from 'json-schema';
import { compile } from 'json-schema-to-typescript';

import {
  GeneratedFileConfig,
  GeneratedFileConfigSchema
} from './GeneratedFile';
import {
  Item,
  ItemConfig,
  DEFAULT_JSON_TO_TS_OPTIONS,
  ItemConfigSchema
} from './Item';
import { renderValueToString, getIndentPrefix } from './ast';
import { resolvePath } from './util';
import { File } from './File';

const MAPPER_TEMPLATE_PATH = path.join(__dirname, 'templates', 'Mapper.ts.ejs');

type DynamoDBIndexableScalarType = 'S' | 'N';
const dynamoDBIndexableScalarTypeMap = new Map<
  JSONSchema4TypeName,
  DynamoDBIndexableScalarType
>([['string', 'S'], ['number', 'N'], ['integer', 'N']]);

interface SecondaryIndex {
  name: string;
  projectionType: string;
  nonKeyAttributes?: string[];
}

const SecondaryIndexSchema = joi
  .object({
    name: joi.string().required(),
    projectionType: joi.string().required(),
    nonKeyAttributes: joi.array().items(joi.string())
  })
  .unknown(false);

interface GlobalSecondaryIndex extends SecondaryIndex {
  hashKey: string;
  rangeKey?: string;
}

const GlobalSecondaryIndexSchema = joi
  .object({
    hashKey: joi.string().required(),
    rangeKey: joi.string()
  })
  .concat(SecondaryIndexSchema)
  .unknown(false);

interface LocalSecondaryIndex extends SecondaryIndex {
  rangeKey: string;
}

const LocalSecondaryIndexSchema = joi
  .object({
    rangeKey: joi.string().required()
  })
  .concat(SecondaryIndexSchema)
  .unknown(false);

interface TableSchemaConfigSchema {
  hashKey: string;
  rangeKey?: string;
  globalSecondaryIndexes?: GlobalSecondaryIndex[];
  localSecondaryIndexes?: LocalSecondaryIndex[];
}

const TableSchemaConfigSchema = joi
  .object({
    hashKey: joi.string().required(),
    rangeKey: joi.string(),
    globalSecondaryIndexes: joi.array().items(GlobalSecondaryIndexSchema),
    localSecondaryIndexes: joi.array().items(LocalSecondaryIndexSchema)
  })
  .unknown(false);

interface TableSchema {
  hashKey: string;
  rangeKey?: string;
  globalSecondaryIndexes: GlobalSecondaryIndex[];
  localSecondaryIndexes: LocalSecondaryIndex[];
}

export interface MapperMetadata extends TableSchemaConfigSchema {
  tableName: string;
}

const MapperMetadataSchema = joi.object({
  tableName: joi.string().required()
});

export interface MapperConfig extends GeneratedFileConfig, MapperMetadata {
  item: ItemConfig;
}

export const MapperConfigSchema = joi
  .object({
    item: ItemConfigSchema.required()
  })
  .concat(MapperMetadataSchema)
  .concat(TableSchemaConfigSchema)
  .concat(GeneratedFileConfigSchema)
  .unknown(false);

export interface MapperRenderProps {
  name: string;
  tableSchema: string;
  generatedPartialMappings: string;
  itemSchema: string;
  partialItemSchema: string;
  keyTypeDeclaration: string;
  keyTypeName: string;
  itemResolvePath: string;
  defaultProviderImports: string;
  indexNameDeclaration: string;
  item: {
    typeName: string;
    uninitializedTypename: string;
  };
}

export class Mapper {
  constructor(name: Readonly<string>, config: Readonly<MapperConfig>) {
    this.name = name;
    this.item = new Item(config.item);
    this.tableName = config.tableName;
    this.tableSchema = {
      hashKey: config.hashKey,
      rangeKey: config.rangeKey,
      globalSecondaryIndexes: config.globalSecondaryIndexes || [],
      localSecondaryIndexes: config.localSecondaryIndexes || []
    };
    this.hashKey = config.hashKey;
    this.rangeKey = config.rangeKey;

    this.generatedFileConfig = {
      outPath: config.outPath,
      overwrite: config.overwrite
    };
  }

  readonly item: Item;
  readonly name: string;
  readonly hashKey: string;
  readonly rangeKey?: string;
  readonly tableName: string;
  readonly tableSchema: TableSchema;
  readonly generatedFileConfig: GeneratedFileConfig;

  private get keyTypeName() {
    return `${this.item.typeName}Key`;
  }

  private resolveIndexKeyType(key: string): string {
    const schemaTypeOrTypes = this.item.resolveKeyType(key);

    if (Array.isArray(schemaTypeOrTypes)) {
      throw new Error('Union types are not currently supported as index keys');
    }

    const schemaType = schemaTypeOrTypes as JSONSchema4TypeName;
    return String(dynamoDBIndexableScalarTypeMap.get(schemaType));
  }

  private renderDefaultProviderImports(): string {
    return (
      Object.values(this.item.defaultProviders)
        .map((defaultProvider) => {
          const modulePath = defaultProvider.relative
            ? resolvePath(
                this.generatedFileConfig.outPath,
                defaultProvider.modulePath
              )
            : defaultProvider.modulePath;
          return defaultProvider.type === 'default'
            ? `import ${defaultProvider.binding} = require('${modulePath}');`
            : `import { ${defaultProvider.binding} } from '${modulePath}';`;
        })
        .join('\n') + '\n'
    );
  }

  private renderGeneratedPartialMappings(): string {
    return (
      Object.entries(this.item.defaultProviders)
        .map(([name, defaultProvider]) => {
          return `${getIndentPrefix(3)}${name}: ${
            defaultProvider.binding
          }(),\n`;
        })
        .join('') + getIndentPrefix(3)
    );
  }

  private renderKeyType(): Promise<string> {
    const { indexKeys } = this;
    const indexProperties = Object.entries(this.item.schema.properties).reduce(
      (acc, [key, schema]) =>
        indexKeys.includes(key) ? { ...acc, [key]: schema } : acc,
      {}
    );

    return compile(
      {
        ...this.item.schema,
        required: [],
        properties: indexProperties
      },
      this.keyTypeName,
      DEFAULT_JSON_TO_TS_OPTIONS
    );
  }

  private renderTableSchema(): string {
    const TableName = this.tableName;
    const KeySchema = [
      {
        AttributeName: this.hashKey,
        KeyType: 'HASH'
      }
    ];

    const AttributeDefinitions = [
      {
        AttributeName: this.hashKey,
        AttributeType: this.resolveIndexKeyType(this.hashKey)
      }
    ];

    if (this.rangeKey) {
      const AttributeName = this.rangeKey;

      KeySchema.push({
        AttributeName,
        KeyType: 'RANGE'
      });

      AttributeDefinitions.push({
        AttributeName,
        AttributeType: this.resolveIndexKeyType(this.rangeKey)
      });
    }

    return renderValueToString(
      {
        TableName,
        KeySchema,
        AttributeDefinitions
      },
      {
        // set offset for property declaration inside of class body
        noPrefixIndent: true,
        indentLevelOffset: 1
      }
    );
  }

  private get itemResolvePath(): string {
    return resolvePath(
      this.generatedFileConfig.outPath,
      this.item.generatedFileConfig.outPath
    );
  }

  private get indexKeys(): string[] {
    return [
      ...this.tableSchema.globalSecondaryIndexes,
      ...this.tableSchema.localSecondaryIndexes,
      { hashKey: this.hashKey, rangeKey: this.rangeKey }
    ].reduce<string[]>((acc, curr: { hashKey?: string; rangeKey?: string }) => {
      if (curr.hashKey) {
        acc.push(curr.hashKey);
      }
      if (curr.rangeKey) {
        acc.push(curr.rangeKey);
      }
      return acc;
    }, []);
  }

  private renderIndexNameType(): string {
    const names = [
      ...this.tableSchema.globalSecondaryIndexes,
      ...this.tableSchema.localSecondaryIndexes
    ].map((index: SecondaryIndex) => index.name);
    const type = names.length
      ? names.map((name) => `"${name}"`).join(' | ')
      : 'never';
    return `export type IndexName = ${type};`;
  }

  private renderItemSchemaValue() {
    return renderValueToString(this.item.schema, {
      noPrefixIndent: true,
      indentLevelOffset: 1
    });
  }

  private renderPartialItemSchemaValue() {
    return renderValueToString(
      {
        ...this.item.schema,
        required: []
      },
      { noPrefixIndent: true, indentLevelOffset: 1 }
    );
  }

  private render(props: MapperRenderProps): Promise<string> {
    return ejs.renderFile(MAPPER_TEMPLATE_PATH, props);
  }

  async generate(): Promise<File[]> {
    const itemFile = await this.item.generate();
    const content = await this.render({
      name: this.name,
      itemResolvePath: this.itemResolvePath,
      tableSchema: this.renderTableSchema(),
      generatedPartialMappings: this.renderGeneratedPartialMappings(),
      itemSchema: this.renderItemSchemaValue(),
      partialItemSchema: this.renderPartialItemSchemaValue(),
      keyTypeDeclaration: await this.renderKeyType(),
      defaultProviderImports: this.renderDefaultProviderImports(),
      indexNameDeclaration: this.renderIndexNameType(),
      keyTypeName: this.keyTypeName,
      item: {
        typeName: this.item.typeName,
        uninitializedTypename: this.item.uninitializedTypeName
      }
    });
    const mapperFile = new File(this.generatedFileConfig.outPath, content);
    return [itemFile, mapperFile];
  }
}
