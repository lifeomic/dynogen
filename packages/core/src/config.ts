import * as joi from '@hapi/joi';

import { MapperConfig, MapperConfigSchema } from './Mapper';

export interface Config {
  mappers: Record<string, MapperConfig>;
}

export const ConfigSchema = joi.object({
  mappers: joi.object().pattern(joi.string(), MapperConfigSchema)
});

export function mustConfig(object: Object): Config {
  joi.assert(object, ConfigSchema, 'Invalid config');
  return object as Config;
}
