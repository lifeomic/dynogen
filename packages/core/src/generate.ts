import { Config } from './config';
import { Mapper } from './Mapper';
import * as fs from 'fs-extra';
import { File } from './File';
import { flatten } from 'lodash';

export async function generateFiles(config: Readonly<Config>): Promise<File[]> {
  const mappers = Object.entries(config.mappers).map(
    ([name, config]) => new Mapper(name, config)
  );

  const files = await Promise.all(mappers.map((mapper) => mapper.generate()));
  return flatten(files);
}

export async function generate(config: Readonly<Config>): Promise<void> {
  const files = await generateFiles(config);
  await Promise.all(
    files.map((file: File) => fs.writeFile(file.path, file.content))
  );
}
