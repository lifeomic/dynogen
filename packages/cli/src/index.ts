#!/usr/bin/env node

import yargs = require('yargs');
import { DEFAULT_DYNOGEN_CONFIG_PATH } from './constants';

import { generateHandler } from './generate';
import { panic } from './util';

function cmd<T>(handler: (context: T) => Promise<void> | void) {
  return (context: T) => {
    try {
      return handler(context);
    } catch (err) {
      panic(err);
    }
  };
}

// tslint:disable-next-line no-unused-expression
yargs
  .scriptName('dynogen')
  .usage('dynogen <cmd> [args]')
  .recommendCommands()
  .command({
    command: 'generate',
    describe: 'Build DynamoDB mappers from config',
    builder: (yargs) =>
      yargs.option('config', {
        type: 'string',
        default: DEFAULT_DYNOGEN_CONFIG_PATH,
        description: 'path to dynogen config'
      }),
    handler: cmd(generateHandler)
  })
  .command({
    command: ['help'],
    handler: () => yargs.showHelp()
  })
  .demandCommand()
  .recommendCommands()
  .strict().argv;
