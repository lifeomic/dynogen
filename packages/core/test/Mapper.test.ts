import { MapperConfig, Mapper } from '../src';
import { createMockMapperConfig } from './helpers/mapper';
import { createMockItemConfig } from './helpers/item';

interface Fixture {
  title: string;
  config: MapperConfig;
  name: string;
}

const fixtures: Fixture[] = [
  {
    title: 'Generates simple Mapper',
    name: 'UserMapper',
    config: createMockMapperConfig()
  },
  {
    title: 'Generates simple Mapper',
    name: 'UserMapper',
    config: createMockMapperConfig({
      rangeKey: 'name',
      globalSecondaryIndexes: [
        {
          name: 'byLogin',
          hashKey: 'login',
          projectionType: 'all'
        }
      ],
      localSecondaryIndexes: [
        {
          name: 'byName',
          rangeKey: 'name',
          projectionType: 'all'
        }
      ],
      item: createMockItemConfig({
        defaultProviders: {
          id: 'uuid=uuid/v4',
          name: './util#generateRandomName'
        }
      })
    })
  }
];

fixtures.forEach(({ title, name, config }) => {
  test(title, async () => {
    const item = new Mapper(name, config);
    const file = await item.generate();
    expect(file).toMatchSnapshot();
  });
});

test('Throws when using union type for index key', async () => {
  const mapper = new Mapper(
    'UserMapper',
    createMockMapperConfig({
      item: createMockItemConfig({
        schema: {
          type: 'object',
          name: 'User',
          properties: {
            id: {
              type: ['string', 'number']
            }
          },
          required: ['id']
        }
      })
    })
  );

  await expect(mapper.generate()).rejects.toThrow(
    'Union types are not currently supported as index keys'
  );
});
