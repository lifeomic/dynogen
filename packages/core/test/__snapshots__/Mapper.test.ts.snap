// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates simple Mapper 1`] = `
Array [
  File {
    "content": "/* tslint:disable */

/**
 * This file was generated with Dynogen.
 * DO NOT EDIT!
 */
export interface User {
  id: string;
  name?: string;
  login: string;
}

export type UninitializedUser = User;

export type GeneratedPartialUser = {};",
    "path": "./User.ts",
  },
  File {
    "content": "/* tslint:disable */

/**
 * This file was generated with Dynogen.
 * DO NOT EDIT!
 */
import { DocumentClient } from 'aws-sdk/clients/dynamodb';
import * as Ajv from 'ajv';

import {
  User,
  UninitializedUser
} from './User';




export interface UserKey {
  id?: string;
}


type IndexName = never;

const ajv = new Ajv({
  allErrors: true
});

export class UserMapper {
  constructor(private readonly client: DocumentClient) {}

  static readonly tableSchema = {
    TableName: \\"Users\\",
    KeySchema: [
      {
        AttributeName: \\"id\\",
        KeyType: \\"HASH\\"
      }
    ],
    AttributeDefinitions: [
      {
        AttributeName: \\"id\\",
        AttributeType: \\"S\\"
      }
    ]
  };

  static get tableName(): string {
    return UserMapper.tableSchema.TableName;
  }

  static initialize(item: UninitializedUser): User {
    return {
      ...item
    };
  }

  private static readonly validator = ajv.compile({
    additionalProperties: false,
    name: \\"User\\",
    type: \\"object\\",
    required: [
      \\"id\\",
      \\"login\\"
    ],
    properties: {
      id: {
        type: \\"string\\"
      },
      name: {
        type: \\"string\\"
      },
      login: {
        type: \\"string\\"
      }
    }
  });
  
  private static readonly partialValidator = ajv.compile({
    additionalProperties: false,
    name: \\"User\\",
    type: \\"object\\",
    required: [

    ],
    properties: {
      id: {
        type: \\"string\\"
      },
      name: {
        type: \\"string\\"
      },
      login: {
        type: \\"string\\"
      }
    }
  });

  static async validate(item: Readonly<User>): Promise<void> {
    if (!await UserMapper.validator(item)) throw new Error(ajv.errorsText());
  }

  static async partialValidate(item: Readonly<Partial<User>>): Promise<void> {
    if (!await UserMapper.partialValidator(item)) throw new Error(ajv.errorsText());
  }

  async put(toPut: Readonly<UninitializedUser>): Promise<User> {
    const item = {
      ...toPut,
      ...UserMapper.initialize(toPut)
    } as User;

    await UserMapper.validate(item);
    await this.client
      .put({
        TableName: UserMapper.tableName,
        Item: item
      })
      .promise();

    // The input item is returned because getting new fields on PutItem
    // operation is impossible. DynamoDB only supports 'OLD_VALUES' and
    // 'NONE' for this operation. This would otherwise require a
    // subsequent query.
    return item;
  }

  async get(key: Readonly<UserKey>): Promise<User> {
    const { Item: item } = await this.client
      .get({
        TableName: UserMapper.tableName,
        Key: key
      })
      .promise();

    if (!item) {
      throw new Error('User not found');
    }

    return item as User;
  }

  async delete(key: Readonly<UserKey>): Promise<void> {
    await this.client
      .delete({
        TableName: UserMapper.tableName,
        Key: key
      })
      .promise();
  }

  async query(): Promise<User[]> {
    const result = await this.client
      .query({
        TableName: UserMapper.tableName
      })
      .promise();

    return result.Items as User[];
  }

  async update(
    key: Readonly<UserKey>,
    patch: Readonly<Partial<User>>
  ): Promise<User> {
    await UserMapper.partialValidate(patch);
    const attributeUpdates = Object.entries(patch).reduce(
      (acc, [key, value]) => ({
        ...acc,
        [key]: {
          Value: value,
          Action: \\"PUT\\"
        }
      }),
      {}
    );

    const { Attributes: old } = await this.client
      .update({
        TableName: UserMapper.tableName,
        Key: key,
        AttributeUpdates: attributeUpdates,
        ReturnValues: \\"ALL_OLD\\"
      })
      .promise();

    return {
      ...(old as User),
      ...patch
    };
  }
}",
    "path": "./Mapper.ts",
  },
]
`;

exports[`Generates simple Mapper 2`] = `
Array [
  File {
    "content": "/* tslint:disable */

/**
 * This file was generated with Dynogen.
 * DO NOT EDIT!
 */
export interface User {
  id: string;
  name?: string;
  login: string;
}

export interface UninitializedUser {
  id?: string;
  name?: string;
  login: string;
}

export interface GeneratedPartialUser {
  id: string;
  name: string;
}",
    "path": "./User.ts",
  },
  File {
    "content": "/* tslint:disable */

/**
 * This file was generated with Dynogen.
 * DO NOT EDIT!
 */
import { DocumentClient } from 'aws-sdk/clients/dynamodb';
import * as Ajv from 'ajv';

import {
  User,
  UninitializedUser
} from './User';

import uuid = require('uuid/v4');
import { generateRandomName } from './util';


export interface UserKey {
  id?: string;
  name?: string;
  login?: string;
}


type IndexName = \\"byLogin\\" | \\"byName\\";

const ajv = new Ajv({
  allErrors: true
});

export class UserMapper {
  constructor(private readonly client: DocumentClient) {}

  static readonly tableSchema = {
    TableName: \\"Users\\",
    KeySchema: [
      {
        AttributeName: \\"id\\",
        KeyType: \\"HASH\\"
      },
      {
        AttributeName: \\"name\\",
        KeyType: \\"RANGE\\"
      }
    ],
    AttributeDefinitions: [
      {
        AttributeName: \\"id\\",
        AttributeType: \\"S\\"
      },
      {
        AttributeName: \\"name\\",
        AttributeType: \\"S\\"
      }
    ]
  };

  static get tableName(): string {
    return UserMapper.tableSchema.TableName;
  }

  static initialize(item: UninitializedUser): User {
    return {
      id: uuid(),
      name: generateRandomName(),
      ...item
    };
  }

  private static readonly validator = ajv.compile({
    additionalProperties: false,
    name: \\"User\\",
    type: \\"object\\",
    required: [
      \\"id\\",
      \\"login\\"
    ],
    properties: {
      id: {
        type: \\"string\\"
      },
      name: {
        type: \\"string\\"
      },
      login: {
        type: \\"string\\"
      }
    }
  });
  
  private static readonly partialValidator = ajv.compile({
    additionalProperties: false,
    name: \\"User\\",
    type: \\"object\\",
    required: [

    ],
    properties: {
      id: {
        type: \\"string\\"
      },
      name: {
        type: \\"string\\"
      },
      login: {
        type: \\"string\\"
      }
    }
  });

  static async validate(item: Readonly<User>): Promise<void> {
    if (!await UserMapper.validator(item)) throw new Error(ajv.errorsText());
  }

  static async partialValidate(item: Readonly<Partial<User>>): Promise<void> {
    if (!await UserMapper.partialValidator(item)) throw new Error(ajv.errorsText());
  }

  async put(toPut: Readonly<UninitializedUser>): Promise<User> {
    const item = {
      ...toPut,
      ...UserMapper.initialize(toPut)
    } as User;

    await UserMapper.validate(item);
    await this.client
      .put({
        TableName: UserMapper.tableName,
        Item: item
      })
      .promise();

    // The input item is returned because getting new fields on PutItem
    // operation is impossible. DynamoDB only supports 'OLD_VALUES' and
    // 'NONE' for this operation. This would otherwise require a
    // subsequent query.
    return item;
  }

  async get(key: Readonly<UserKey>): Promise<User> {
    const { Item: item } = await this.client
      .get({
        TableName: UserMapper.tableName,
        Key: key
      })
      .promise();

    if (!item) {
      throw new Error('User not found');
    }

    return item as User;
  }

  async delete(key: Readonly<UserKey>): Promise<void> {
    await this.client
      .delete({
        TableName: UserMapper.tableName,
        Key: key
      })
      .promise();
  }

  async query(): Promise<User[]> {
    const result = await this.client
      .query({
        TableName: UserMapper.tableName
      })
      .promise();

    return result.Items as User[];
  }

  async update(
    key: Readonly<UserKey>,
    patch: Readonly<Partial<User>>
  ): Promise<User> {
    await UserMapper.partialValidate(patch);
    const attributeUpdates = Object.entries(patch).reduce(
      (acc, [key, value]) => ({
        ...acc,
        [key]: {
          Value: value,
          Action: \\"PUT\\"
        }
      }),
      {}
    );

    const { Attributes: old } = await this.client
      .update({
        TableName: UserMapper.tableName,
        Key: key,
        AttributeUpdates: attributeUpdates,
        ReturnValues: \\"ALL_OLD\\"
      })
      .promise();

    return {
      ...(old as User),
      ...patch
    };
  }
}",
    "path": "./Mapper.ts",
  },
]
`;
